File: node-bindings/classic-node/__test__/update_rejection_conformance.ts
import * as classic from "../index";

/** Exercise original default helpers using a synthetic header that reqwest rejects before transport. */
export async function observeUpdateRejection(fixture: Record<string, any>): Promise<Record<string, any>> {
    if (fixture.operation !== "latest" || fixture.token !== "synthetic\ninvalid") throw new Error("Unsupported rejection fixture");
    const previous = process.env.GITHUB_TOKEN;
    process.env.GITHUB_TOKEN = fixture.token;
    try {
        for (const invoke of [
            () => classic.getLatestRelease(fixture.owner, fixture.repo),
            () => classic.checkForUpdates(fixture.owner, fixture.repo, fixture.currentVersion),
        ]) {
            let rejected = false;
            try {
                await invoke();
            } catch (error) {
                if (!(error instanceof Error) || error.message !== "HTTP error: builder error") throw error;
                rejected = true;
            }
            if (!rejected) throw new Error("Malformed authorization header was accepted");
        }
        return {boundary: "request-builder", error: "builder-error", requestBuilt: false};
    } finally {
        // Keep synthetic credentials scoped until both asynchronous client calls finish.
        if (previous === undefined) delete process.env.GITHUB_TOKEN;
        else process.env.GITHUB_TOKEN = previous;
    }
}

File: python-bindings/tests/update_rejection_conformance.py
"""Exercise default public update APIs only at deterministic pre-transport failure boundaries."""

import asyncio


def observe_update_rejection(fixture):
    """Require typed installed-version rejection or the native malformed-header builder error."""
    import classic_update

    if fixture["token"] != "synthetic\ninvalid":
        raise ValueError("only the synthetic invalid credential is permitted")
    if fixture["operation"] == "metadata":
        client = classic_update.GithubClient(
            fixture["owner"], fixture["repo"], fixture["token"]
        )
        return {"repoUrl": client.repo_url()}
    if fixture["operation"] == "notification":
        try:
            classic_update.check_app_notification(
                fixture["owner"], fixture["repo"], fixture["invalidVersion"]
            )
        except classic_update.ClassicNotificationInstalledVersionParseError:
            return {
                "boundary": "caller-validation",
                "error": "invalid-installed-version",
                "requestBuilt": False,
            }
        raise ValueError("invalid installed version was accepted")

    async def reject():
        """Await the real client; transport errors and successful requests are not accepted evidence."""
        client = classic_update.GithubClient(
            fixture["owner"], fixture["repo"], fixture["token"]
        )
        try:
            if fixture["operation"] == "latest":
                await client.get_latest_release()
            elif fixture["operation"] == "all":
                await client.get_all_releases(False, False)
            else:
                raise ValueError("unsupported rejection operation")
        except RuntimeError as error:
            if str(error) != "GitHub API error: HTTP error: builder error":
                raise
            return {
                "boundary": "request-builder",
                "error": "builder-error",
                "requestBuilt": False,
            }
        raise ValueError("malformed authorization header was accepted")

    return asyncio.run(reject())
